### PR TITLE
Feature: detailed tag view and portfolios

### PR DESCRIPTION
**Pull request by matusziak**
Addresses #58 and  #59

image
image
image

### CODE REVIEW

**Comment by Duskyman** [comment]
It may not be probably related to changes in this PR, but `api/tags` returned duplicate tags from my db. Running the query directly did the same. I guess there is some kind of error in the query.
Apart from that, we found while calling that the newer messages without states "overwrite" any state that was there before.

I compiled a little bit verbose, but I believe more readable and easier to debug query(except the t1 - t4 subquery names 😅) that seems to return correct results. **NOTE: it's not 100% complete**:
```SQL
-- collect all data together
select t1.last_report_ts, tagged.tag, t3.state, tag.*, portfolios from tagged 
JOIN (
	-- get the most recent reports
	select tag, max(report) last_report_ts 
	from tagged 
	group by tag
) t1 ON (tagged.report = t1.last_report_ts and tagged.tag = t1.tag)
LEFT JOIN (
	-- get the most recent states
	select tagged.tag, tagged.state from tagged JOIN (
		-- get the most recent tags containing state
		select tag, max(report) last_state_ts 
			from (
				-- find all tags with states
				select tag, report 
				from tagged 
				where state is not null
			) t1
			group by tag
	) t2 ON (t2.last_state_ts = tagged.report and t2.tag = tagged.tag)
) t3 ON (t3.tag = tagged.tag)
-- add tag info to the collection
JOIN tag on (tagged.tag = tag.tag)
LEFT JOIN (
	-- add portfolios to the collection
	select tag, array_agg(portfolio) as portfolios 
	from tag_portfolio 
	group by tag
) t4
ON (tag.tag = t4.tag)
```

File: server/src/db.js [hunk 1]
```diff
@@ -1,34 +1,112 @@
...
+            FROM (SELECT tag, Count(*) AS count, Max(report) AS lastTs
+                  FROM   tagged 
+                  WHERE  NOT is_command 
+                  GROUP  BY tag) t1 
+            INNER JOIN tagged t2 ON t1.lastTs=t2.report) t 
+      NATURAL JOIN tag tg
+      LEFT JOIN (SELECT tag AS tag, array_agg(portfolio) AS portfolios 
+      FROM tag_portfolio 
+      GROUP BY tag) p USING (tag) 
+      ORDER BY tag ASC`
```
Comment (Duskyman) [inline server/src/db.js]: I wrote a suggestion regarding this query in a separate comment

File: client/src/components/message/MessageList.js [hunk 2]
```diff
@@ -27,9 +27,9 @@ class MessageList extends Component {
   render () {
     return (
       <div className="MessageList">
-        {this.props.reports.map (report => (
+        {this.props.reports.map ((report, i) => (
           <MessageItem
-            key={report.id}
+            key={i}
```
Comment (Duskyman) [inline client/src/components/message/MessageList.js:32]: Is this change required? Are `report.id` not unique?

File: client/src/components/leftPanel/TagEdit.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    portfolios: this.props.tag.portfolios?.map(
+      (p) => ({ label: p, value: p })
+    ) || [],
+    isArchived: this.props.tag.isArchived || false,
+    loading: false,
+    resultType: '',
+  }
+
+  onClose = () => {
+    if (this.props.history.length > 2) {
```
Comment (Duskyman) [inline client/src/components/leftPanel/TagEdit.js]: `this.props.history.length > 2` implies we want at least 3 pages in history. Is this wanted? The first one is about:blank / default browser page?

File: client/src/components/leftPanel/TagEdit.js [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    const data = {
+      tag: tag.tag,
+      isArchived,
+      asanaLink,
+      portfolios: portfolios.map((p) => p.value),
+    }
+
+    this.setState({ loading: true })
+    const result = await updateTag(data)
+    await this.props.loadReports() // refresh tag list after update
```
Comment (Duskyman) [inline client/src/components/leftPanel/TagEdit.js]: You are already destructuring props few lines above. Move this up there as well :)

File: client/src/components/leftPanel/LeftPanel.js [hunk 5]
```diff
@@ -1,40 +1,84 @@
...
+  toggleDetailedView = () => this.setState({ detailedView: !this.state.detailedView })
 
   render() {
-    const {tags} = this.props
+    const {tags, users, reports, portfolioOptions} = this.props
     for (const tag of tags) {
       tag.isLate =
         !tag.isArchived && moment().diff(parseTs(tag.lastTs), 'days') > tag.frequency
     }
     const partitionedTags = _.flatten(_.partition(tags, 'isLate'))
```
Comment (Duskyman) [inline client/src/components/leftPanel/LeftPanel.js:54]: Why do we partition something into two arrays and then immediately flatten it into one array? For ordering?

File: client/src/components/leftPanel/EditPortfolios.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    this.props.history.push(`/`)
+  }
+
+  onAddInputChange = (e) => {
+    this.setState({resultType: ''}) // reset success/error
+    this.setState({addInput: e.target.value})
+  }
+
+  onCreate = async (e, name) => {
+    e.preventDefault() // prevent reloading page with button click
```
Comment (Duskyman) [inline client/src/components/leftPanel/EditPortfolios.js:25]: I believe this is default behaviour on `<input type="submit" />`, not on `<button />` element.

File: client/src/components/leftPanel/EditPortfolios.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+class EditPortfolios extends Component {
+  state = {
+    addInput: '',
+    deleteSearchInput: '',
+    loading: false,
+    resultType: ''
+  }
+
+  onClose = () => {
+    this.props.history.push(`/`)
```
Comment (Duskyman) [inline client/src/components/leftPanel/EditPortfolios.js]: Maybe you want to do the same check as in `TagEdit`. Or maybe not 🤔 
```javascript
if (this.props.history.length > 2) {
  this.props.history.goBack()
} else {
  // in case tag edit was first visited site
  this.props.history.push(`/${encodeURI(this.props.tag.tag)}`)
}
```

File: client/src/App.js [hunk 8]
```diff
@@ -16,27 +18,58 @@ class App extends Component {
...
 
+  loadPortfolioOptions = async () => {
+    const options = await getPortfolios()
+    this.setState({ portfolioOptions: options.map((p) => ({label: p.name, value: p.name})) })
+  }
+
   async componentDidMount() {
     await this.loadReports()
     this.refreshInterval = setInterval(this.loadReports, 5 * 60 * 1000)
+    await this.loadPortfolioOptions()
```
Comment (Duskyman) [inline client/src/App.js]: Little performance boost suggestion: await Promise.all() containing these two async calls.

**Review by Duskyman — CHANGES_REQUESTED** [review]
Fix the db bugs and check the suggestions. It looks and works fine otherwise 😄 

image

File: client/src/components/leftPanel/TagEdit.js (hunk 3 above)
Comment (matusziak) [reply client/src/components/leftPanel/TagEdit.js]: Yeah, correct me if I'm wrong. The reasoning is first you get the default, if you then copy the url to edit tag directly, then on modal close we want to go to the tags page with previous tag selected. If we got to this tag edit page from the tags page, then we want to "go back" there to preserve url parameters such as portfolios filter.

File: client/src/components/leftPanel/LeftPanel.js (hunk 5 above)
Comment (matusziak) [reply client/src/components/leftPanel/LeftPanel.js:54]: I believe this orders the tags in a way, which allows to see "late" tags first (if they exceed set frequency by late reporting).

File: client/src/components/leftPanel/EditPortfolios.js (hunk 7 above)
Comment (matusziak) [reply client/src/components/leftPanel/EditPortfolios.js]: I was thinking that this will not be used too often to need it behave in more intelligent way, and if you delete some portfolios it was easier just to reset the page, rather than checking portfolios in the url filter parameters.

File: client/src/components/message/MessageList.js (hunk 2 above)
Comment (matusziak) [reply client/src/components/message/MessageList.js:32]: There were no ids on the report objects. Maybe this was caused by some backend previous changes which omitted the ids...but when I opened the app this error needed to be solved.

File: client/src/components/leftPanel/LeftPanel.js (hunk 5 above)
Comment (Duskyman) [reply client/src/components/leftPanel/LeftPanel.js:54]: I see it now - late tags are marked and then the partition takes place, which puts late tags into first array and rest to the second. Then they are flattened

File: client/src/components/leftPanel/EditPortfolios.js (hunk 6 above)
Comment (Duskyman) [reply client/src/components/leftPanel/EditPortfolios.js:25]: Therefore you can remove this line, as it has no effect for <button /> tag (they don't reload the page)

File: server/src/db.js [hunk 9]
```diff
@@ -31,19 +31,40 @@ export async function loadReplies() {
...
+      ) t3 ON (t3.tag = tagged.tag)
+      -- add tag info to the collection
+      JOIN tag on (tagged.tag = tag.tag)
+      LEFT JOIN (
+        -- add portfolios to the collection
+        SELECT tag, array_agg(portfolio) AS portfolios 
+        FROM tag_portfolio 
+        group by tag
+      ) t4
+      ON (tag.tag = t4.tag)`
```
Comment (Duskyman) [inline server/src/db.js:67]: I would `ORDER BY tag ASC` the results like in previous solution to make sure they are in alphabetic order.

**Review by Duskyman — APPROVED** [review]
One more little suggestion regarding the query in the comment. Have you checked that the results are valid? I don't write queries that often 😅, no warranties are included.

File: server/src/db.js (hunk 9 above)
Comment (matusziak) [reply server/src/db.js:67]: I think the ordering logic is more complex and it is done on the FE anyway.

File: client/src/components/leftPanel/LeftPanel.js [hunk 10]
```diff
@@ -86,10 +108,20 @@ class LeftPanel extends Component {
...
           )}
+          {filteredTags.map((tag) => (isTagEditOpen[tag.tag] && <TagEdit
+            key={tag.tag}
+            tag={tag}
+            onClose={this.onCloseTagEdit}
+            isArchived={tag.isArchived}
+            asanaLink={tag.asanaLink}
+            portfolioOptions={portfolioOptions}
+            loadReports={loadReports}
+          />))}
```
Comment (Duskyman) [inline client/src/components/leftPanel/LeftPanel.js]: This does not seem right 🤔. We can edit only one tag at time and thus we should store in state only one tag - the currently edited one (or null if none is being edited).

**Review by Duskyman — CHANGES_REQUESTED** [review]
One more change request. Navigation works like a charm, however we lost ability to send links directly to edits / portfolios, but I believe that is not big deal.
